src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -202,4 +209,71 @@ public static double calculateTPCMetrics(
 			return Double.POSITIVE_INFINITY;
 		}
 	}
+
+	/**
+	 *
+	 * @param dataModel
+	 * @param obsoletionPeriodMs the maximum amount of time (in milliseconds) it is
+	 *                           worth aggregating over starting from the most
+	 *                           recent scan entry and working backwards in time
+	 * @param agg                an aggregator to output a calculated "aggregated
+	 *                           signal strength"
+	 * @return a map from BSSID to an "aggregated wifiscan entry" (i.e., a single
+	 *         entry with its {@code signal} attribute modified to be the aggregated
+	 *         signal value instead of the value in just the most recent entry.
+	 */
+	public Map<String, WifiScanEntry> getAggregatedWifiScans(Modeler.DataModel dataModel, long obsoletionPeriodMs,
+			Aggregator<Double> agg) {
+		/*
+		 * NOTE: if a BSSID does not have a non-obsolete entry, it will be returned
+		 * (i.e., it will not be a key in the returned map).
+		 */
+		Map<String, WifiScanEntry> aggregatedWifiScans = new HashMap<>();
+		for (Map.Entry<String, List<List<WifiScanEntry>>> mapEntry : dataModel.latestWifiScans.entrySet()) {
+			// Flatten the wifiscan entries and sort in reverse chronological order
+			List<List<WifiScanEntry>> scans = mapEntry.getValue();
+			if (scans.isEmpty()) {
+				continue;
+			}
+			List<WifiScanEntry> mostRecentToOldest = scans.stream().flatMap(list -> list.stream())
+					.sorted((entry1, entry2) -> {
+						return -Long.compare(entry1.unixTimeMs, entry2.unixTimeMs);
+					}).collect(Collectors.toUnmodifiableList());
+
+			/*
+			 * For a given BSSID, discard the obsolete entries. Then,consider the most
+			 * recent entry. Take its ht_oper and vht_oper. Discard earlier entries with a
+			 * different ht_oper or a different vht_oper. For earlier entries with the same
+			 * ht_oper and vht_oper, take the average signal value.
+			 */
+			String bssid = mapEntry.getKey();
+			long now = Instant.now().getEpochSecond();
+			String newestHtOper = null;
+			String newestVhtOper = null;
+			for (WifiScanEntry entry : mostRecentToOldest) {
+				if (now - entry.unixTimeMs >= obsoletionPeriodMs) {
+					// discard obsolete entries
+					break;
+				} else if (entry.ht_oper == null && entry.vht_oper == null) {
+					continue;
+				}
+				if (newestHtOper == null) {
+					// may still be null after this assignment
+					newestHtOper = entry.ht_oper;
+				}
+				if (newestVhtOper == null) {
+					// may still be null after this assignment
+					newestVhtOper = entry.vht_oper;
+				}
+				// if the entry matches ht_oper and vht_oper, add its signal to the aggregate
+				if ((entry.ht_oper == null || entry.ht_oper == newestHtOper)
+						&& (entry.vht_oper == null || entry.vht_oper == newestVhtOper)) {
+					aggregatedWifiScans.putIfAbsent(bssid, entry);
+					agg.addValue((double) entry.signal);
+				}
+			}
+			aggregatedWifiScans.get(bssid).signal = (int) Math.round(agg.getAggregate());
```

Comment:
will this aggregate across `bssid`s -- I thought that each one needed to be separately aggregated. Do we need to reset agg somewhere?

Reply:
Good catch.